[nanobind] Add hook to extend ndarray implicit conversions

We also revert upstream 146edeb9fc196ffd0dd2c6fd19cb7c32c1d0986e to
make the patch easier to write.

In theory, this could be upstreamed, but acceptance is doubtful. For a
pertinent discussion, see:
https://github.com/wjakob/nanobind/discussions/327.

XXX porting Instead of hooking nanobind to call into our type-caster of last
resort, we could simply not include nanobind/eigen/dense.h and instead write our
own type_caster that behaves the way we want it to.  That would be better
long-term, but I'm undecided about whether we do it before or after introducing
nanobind support to our master branch.

--- include/nanobind/nb_lib.h
+++ include/nanobind/nb_lib.h
@@ -480,6 +480,15 @@ NB_CORE ndarray_handle *ndarray_import(PyObject *o,
                                        bool convert,
                                        cleanup_list *cleanup) noexcept;
 
+// Hook for additional ndarray implicit conversion.
+NB_INLINE ndarray_handle *ndarray_extra_import(PyObject *,
+                                               const ndarray_config *,
+                                               bool,
+                                               cleanup_list *,
+                                               void*) noexcept {
+  return {};
+}
+
 // Describe a local ndarray object using a DLPack capsule
 NB_CORE ndarray_handle *ndarray_create(void *data, size_t ndim,
                                        const size_t *shape, PyObject *owner,

--- include/nanobind/ndarray.h
+++ include/nanobind/ndarray.h
@@ -574,15 +574,17 @@ template <typename... Args> struct type_caster<ndarray<Args...>> {
             (void) shape_buf;
         }
 
-        detail::ndarray_handle *h = ndarray_import(
-            src.ptr(), &config, flags & (uint8_t) cast_flags::convert, cleanup);
-
-        if (NB_UNLIKELY(value.m_handle))
-            detail::ndarray_dec_ref(value.m_handle);
-        if (NB_LIKELY(h))
-            value.m_dltensor = *detail::ndarray_inc_ref(h);
-        value.m_handle = h;
-        return h != nullptr;
+        value = Value(ndarray_import(src.ptr(), &config,
+                                     flags & (uint8_t) cast_flags::convert,
+                                     cleanup));
+
+        if (value.is_valid()) { return true; }
+
+        value = Value(ndarray_extra_import(src.ptr(), &config,
+                                     flags & (uint8_t) cast_flags::convert,
+                                     cleanup, (Config*)nullptr));
+
+        return value.is_valid();
     }
 
     static handle from_cpp(const ndarray<Args...> &tensor, rv_policy policy,

--- BUILD.bazel
+++ BUILD.bazel
@@ -79,7 +79,7 @@ cc_library(
     visibility = ["//visibility:public"],
     deps = [
         ":nanobind",
-        "@eigen",
+        "@drake//tools/workspace/eigen",
     ],
 )
 

--- MODULE.bazel
+++ MODULE.bazel
@@ -5,6 +5,8 @@ module(
     compatibility_level = 1,
 )
 
+bazel_dep(name = "drake")
+
 bazel_dep(name = "eigen", version = "3.4.0.bcr.3")
 bazel_dep(name = "platforms", version = "0.0.10")
 bazel_dep(name = "robin-map", version = "1.4.1")
